scripts/install-cni.sh

```diff
@@ -130,20 +130,149 @@ else
+  local ip=$1
+
+  [[ "${ip:-}" =~ ${IPV6_VALIDATION_REGEXP} ]]
+}
+
+#######################################
+# Replaces `@subnets` and `@routes` placeholders using `.spec.podCIDRs` from node json.
+# In directpath use case it additionally adds IPv6 subnet derived from node's IPv6 address.
+# Arguments:
+#   node json from kube-apiserver
+#   node_ipv6_addr node's IPv6 address from GCE metadata server
+#######################################
+function fillSubnetsInCniSpecV2Template {
+  local node=$1
+  local node_ipv6_addr=$2
+
+  local SUBNETS_REPLACEMENT=()
+  local ROUTES_REPLACEMENT=()
+
+  local ipv6_subnet_configured="false"
+
+  for subnet in $(jq -r '.spec.podCIDRs[]' <<<"$node") ; do
+    if is_ipv4_range "$subnet" ; then
+      SUBNETS_REPLACEMENT+=("[{\"subnet\": \"${subnet}\"}]")
+      ROUTES_REPLACEMENT+=("{\"dst\": \"0.0.0.0/0\"}")
+    elif is_ipv6_range "$subnet" ; then
+      POPULATE_IP6TABLES="true"
+      ipv6_subnet_configured="true"
+      SUBNETS_REPLACEMENT+=("[{\"subnet\": \"${subnet}\"}]")
+      ROUTES_REPLACEMENT+=("{\"dst\": \"::/0\"}")
+    else
+      echo "[ERROR] Subnet $subnet is not a valid IP range"
+      exit 1
+    fi
+  done
+
+  # Directpath use case
+  if [ "$ENABLE_IPV6" == "true" ] && [ "$ipv6_subnet_configured" == "false" ] ; then
+    # Directpath adds IPv6 subnet and route derived from host with fixed range
+    # of /112 even when it is not specified in node's .spec.podCIDRs
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      POPULATE_IP6TABLES="true"
+      SUBNETS_REPLACEMENT+=("[{\"subnet\": \"${node_ipv6_addr}/112\"}]")
+      ROUTES_REPLACEMENT+=("${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}")
+    fi
+  fi
+
+  SUBNETS_REPLACEMENT_CONCATENATED=$(IFS=', ' ; echo "${SUBNETS_REPLACEMENT[*]}")
+  ROUTES_REPLACEMENT_CONCATENATED=$(IFS=', ' ; echo "${ROUTES_REPLACEMENT[*]}")
+
+  cni_spec=${cni_spec//@subnets/$SUBNETS_REPLACEMENT_CONCATENATED}
+  cni_spec=${cni_spec//@routes/$ROUTES_REPLACEMENT_CONCATENATED}
+}
+
+#######################################
+# Replaces `@ipv4Subnet', '@ipv6SubnetOptional` and `@ipv6RouteOptional` placeholders using `.spec.podCIDR` from node json and node's ipv6 from metadata server.
+# Arguments:
+#   node json from kube-apiserver
+#   node_ipv6_addr node's IPv6 address from GCE metadata server
+#######################################
+function fillSubnetsInCniSpecLegacyTemplate {
+  local node=$1
+  local node_ipv6_addr=$2
+
+  local primary_subnet
+  primary_subnet=$(jq -r '.spec.podCIDR' <<<"$response")
+
+  if is_ipv4_range "${primary_subnet:-}" ; then
+    echo "PodCIDR validation succeeded: ${primary_subnet:-}"
```

Comment:
I'm okay with not keeping the exact old messages. Feel free to make changes if you want.

Reply:
rewritten to better reflect current logic